tidy3d/components/data_new.py

```diff
@@ -0,0 +1,137 @@
+from typing import Dict, List, Union
+import json
+
+import pydantic as pd
+import h5py
+import xarray as xr
+
+from .base import Tidy3dBaseModel
+from .types import ArrayLike
+from .simulation import Simulation
+from .monitor import FieldMonitor, FieldTimeMonitor, FluxMonitor, FluxTimeMonitor, ModeMonitor
+from .monitor import PermittivityMonitor, ModeFieldMonitor
+
+# every type that can be used as a component
+ComponentType = Union[
+    Simulation,
+    FieldMonitor,
+    FieldTimeMonitor,
+    FluxMonitor,
+    FluxTimeMonitor,
+    ModeMonitor,
+    ModeFieldMonitor,
+]
+
+
+class Tidy3dData(Tidy3dBaseModel):
+
+    component: ComponentType = pd.Field(
+        None,
+        title="Original component",
+        description="Copy of the tidy3d component that this data is associated with.",
+    )
+
+    data: xr.DataArray = pd.Field(
+        None,
+        title="Data",
+        description="An xarray DataArray corresponding to the data stord by this element.",
+    )
+
+    data_dict: Dict[str, "Tidy3dData"] = pd.Field(
+        {},
+        title="Data dictionary",
+        description="Collection of other tidy3d data stored by this element",
+    )
```

Comment:
I am wondering if there can be just one field, 

```
    data: Dict[str, Union[xr.DataArray, "Tidy3dData"] = pd.Field(
        {},
        title="Data dictionary",
        description="Collection of other tidy3d data stored by this element",
    )
```

Reply:
So, do we imagine there to be a situation where a component has a data and a data_dict?  if that's not out of the question, I kind of like them separated so we limit the amount of checking needed. For example

```python
if isistnace(self.data, dict):
    do dict stuff
else:
    do xarray stuff

```

Reply:
What I imagine is that you keep nesting until you always reach an xarray. Instances that have a simple data in your formulation would essentially have to associate that to a key. As an example, say you have a `SimulationData` with a `FluxMonitor` (single array data) and a `FieldMonitor` (composite data). In your formulation, getting the corresponding xarrays would be something like

```
flux = sim_data["flux_mnt"].data
Ex = sim_data["field_mnt"]["Ex"].data
```

And what I'm proposing would be

```
flux = sim_data["flux_mnt"]["flux"]
Ex = sim_data["field_mnt"]["Ex"]
```

You also wouldn't need all the `values`, `coords`, `dims` wrappers because you never need to explicitly call `data`. In the current formulation it may also be confusing e.g. when `FieldData` has no `data` but only a `data_dict`. Someone doing `field_data.values` would get `None`. Conceptually I like what I'm proposing more, just not sure if there are any pitfalls.